ocaml/typing/typecore.ml

```diff
@@ -6512,18 +6539,10 @@ and type_application env app_loc expected_mode (position, maybe_outer_ret_mode)
       in
       submode ~loc:app_loc ~env ~reason:(Application ty_ret)
         mode_ret expected_mode;
-      (match maybe_outer_ret_mode with
-      | Some outer_ret_mode -> begin
-        (* This application is at the tail of a function with a region;
-            if ap_mode is local, funct_ret_mode needs to be local as well. *)
-          match
-            Value_mode.submode (Value_mode.of_alloc ap_mode) outer_ret_mode
-          with
-          | Ok () -> ()
-          | Error _ -> raise (Error (app_loc, env, Tail_call_local_returning))
-        end
-      | _ -> ()
-      );
+
+      if not partial_app then
+        check_tail_call_local_returning ap_mode pm.region_mode
+          (Error (app_loc, env, Tail_call_local_returning));
```

Comment:
It would be better to pass the `app_loc` and `env` to the function and have it create the error if needed. It would also be better to pass in the whole `pm`.